### ISSUE TITLE
C++11 idiom

### ISSUE THREAD

**Issue by jjcasmar** [stevengj/nlopt#219]
I found nlopt a great library, but using it through the C++ is really frustrating.
You need to provide a `void*` for passing data to the objective/constraint functions, with all the problemas that may have. 

Also, you need to pass a function pointer, so you cant use lambda functions with captures (that would avoid passing the void*).

Ive written a thin wrapper on top of NLopt which tries to offer a more modern c++ API, enabling the use of lambdas and hiding from the API the use of void*. 

Would you be interested in this to be merge?

**Comment by stevengj** [comment]
Overloading `set_min_objective` etcetera to provide a method that takes a `lambda` sounds good (i.e. of type `ClosureType`).  You could protect it with `#if __cplusplus >= 201103L` so that it doesn't break earlier C++ versions.

**Comment by jjcasmar** [comment]
Its not using overriding at all. I provided a complete new wrapper around the C interface. Actually its a wrapper around the current C++ interface, but as its not complete (preconds), Im going to modify it to be a c++11 wrapper around C. 

It adds some more things as enabling compatibility with Eigen. Instead of forcing the vector type to std::vector<double>, you can specify your own VectorType. The wrapper is in charge of converting it to/from double with some specialized functions. 

Is that okay?

**Comment by stevengj** [comment]
Providing your own wrapper type is fine for an external library, but if you are modifying NLopt itself you should just directly add methods to `class opt`, no?

**Comment by jjcasmar** [comment]
Im not sure if I want to add new functionality to current C++ wrapper. I have already talk about the things I think are not good in the API and which Im trying to improve. 

Thats why I have been thinking on adding a new c++11 wrapper which doesnt interfere with the current implementation. Other option would be to add this new functionality in the current opt class, but I think that would mean having a big class which lots of #ifdef providing two different c++ wrapper. 

Having a user defined VectorType is okay or do you prefer to leave that outside and continue using std::vector<double> only?

Let me think how can I do something that is okay to merge upstream.

**Comment by stevengj** [comment]
>  having a big class which lots of #ifdef providing two different c++ wrapper

It's already pretty big, I don't see a problem with adding a few more methods.  I would prefer to use `std::vector` only.

**Comment by dburov190** [comment]
@stevengj Has this seen any traction since 2019? If not, I could pick it up. I've hit the same problem: I'd like to pass a functor (a class with `operator()` defined) instead of a function pointer so that I could avoid `(void*)` and `(MyClass*)` casts.

I'm thinking of something simple in a style similar to [`boost::math::tools::brent_find_minima`](https://www.boost.org/doc/libs/1_75_0/boost/math/tools/minima.hpp):
```c++
template<typename Functor>
void nlopt::opt::set_max_objective(Functor f);
```

This would allow lambdas with captures and functors to be passed. Unfortunately, I don't think this can be implemented as a thin wrapper around existing `set_max_objective` methods (or it would be insanely inefficient).

**Comment by dburov190** [comment]
@stevengj Hi, just checking if you saw the previous comment? thank you!

**Comment by stevengj** [comment]
> Unfortunately, I don't think this can be implemented as a thin wrapper around existing set_max_objective methods (or it would be insanely inefficient).

Why not?

**Comment by dburov190** [comment]
I think it's impossible to implement that way because either of the variants requires a function of certain signature to be passed as an argument (or, rather, a function pointer):
```c++
typedef double (*vfunc)(const std::vector<double> &x,
			  std::vector<double> &grad, void *data);
void set_max_objective(vfunc vf, void *f_data);
```

So, if I pass to the hypothetical wrapper `set_max_objective(Functor functor)` an object `functor` which has `operator()` implemented, then the wrapper's job is to strip the `operator()` from `functor` and pass it to one of the existing `set_max_objective` functions. However, if I choose to pass `functor.operator()`, this will not match the signature since [any non-static method has `this`](https://stackoverflow.com/questions/41667653/does-every-c-member-function-take-this-as-an-input-implicitly) as an implicit parameter (also [here](https://stackoverflow.com/questions/47718272/pointer-to-a-class-method-in-c)). Besides, [the type of a member-pointer is different](https://stackoverflow.com/questions/1485983/calling-c-class-methods-via-a-function-pointer) from the plain function-pointer.

I could, however, write a different wrapper which would be similar in style to what you have in the library now, i.e. use a custom-defined `struct` and call the C function directly:
```c++
void set_max_objective(Functor functor) {
  another_func_data *d = new another_func_data;
  if (!d) throw std::bad_alloc();
  d->o = this; d->functor = functor;
  d->munge_destroy = d->munge_copy = NULL;
  mythrow(nlopt_set_max_objective(o, myfunctor, d)); // d freed via o
  alloc_tmp();
}
```
(something along the lines, this is not precise yet).

Alternatively, I could add a field to `myfunc_data` to store a pointer to an object instead of introducing `another_func_data`.

**Comment by dburov190** [comment]
Yes, yes! this is exactly what I had in mind, in the code snippet above where I wrote
```c++
mythrow(nlopt_set_max_objective(o, myfunctor, d)); // d freed via o
```
`myfunctor` is (almost) exactly your `functor_wrapper` (: I just tried to follow the style of the library.

You're right about the `std::vector` argument, but I can write two overloaded wrappers that take either `double *` or `std::vector`.

Finally, I'm afraid I'd have to use a template function, something like:
```c++
template <typename Functor>
double functor_wrapper(const std::vector<double> &x, std::vector<double> &grad, void *functor_data);
```
but I guess it's not a big deal since you have everything in `.hpp` anyway, right? Alternatively, I could use `std::function` but that's a bit ugly and cumbersome.
